src/expr/src/typecheck.rs

```diff
@@ -0,0 +1,777 @@
+    },
+    BadTopKGroupKey {
+        source: &'a MirRelationExpr,
+        key: usize,
+        input_type: Vec<ColumnType>,
+    },
+    BadTopKOrdering {
+        source: &'a MirRelationExpr,
+        order: ColumnOrder,
+        input_type: Vec<ColumnType>,
+    },
+    BadLetRecBindings {
+        source: &'a MirRelationExpr,
+    },
+}
+
+type Ctx = BTreeMap<Id, Vec<ColumnType>>;
+
+/// Returns true when it is safe to treat a `got` row as an `expected` row
+///
+/// In particular, the core types must be equal, and if a column in `known` is nullable, that column should also be nullable in `got`
+/// Conversely, it is okay to treat a known non-nullable column as nullable: `got` may be nullable when `known` is not
+pub fn columns_match(got: &[ColumnType], known: &[ColumnType]) -> bool {
+    if got.len() != known.len() {
+        return false;
+    }
+
+    got.iter().zip_eq(known.iter()).all(|(got, known)| {
+        (!known.nullable || got.nullable) && got.scalar_type.base_eq(&known.scalar_type)
+    })
+}
+
+impl MirRelationExpr {
+    /// Returns the type of a relation expression or a type error.
+    ///
+    /// This function is careful to check validity, not just find out the type.
+    ///
+    /// It should be linear in the size of the AST.
+    ///
+    /// ??? should we also compute keys and return a `RelationType`?
+    pub fn typecheck(&self, ctx: &Ctx) -> Result<Vec<ColumnType>, TypeError> {
+        use MirRelationExpr::*;
+
+        match self {
+            Constant { typ, rows } => {
+                if let Ok(rows) = rows {
+                    for (row, _id) in rows {
+                        let datums = row.unpack();
+
+                        // correct length
+                        if datums.len() != typ.column_types.len() {
+                            return Err(TypeError::BadConstantRow {
+                                source: self,
+                                got: row.clone(),
+                                expected: typ.column_types.clone(),
+                            });
+                        }
+
+                        // correct types
+                        if datums
+                            .iter()
+                            .zip_eq(typ.column_types.iter())
+                            .any(|(d, ty)| d != &mz_repr::Datum::Dummy && !d.is_instance_of(ty))
+                        {
+                            return Err(TypeError::BadConstantRow {
+                                source: self,
+                                got: row.clone(),
+                                expected: typ.column_types.clone(),
+                            });
+                        }
+                    }
+                }
+
+                Ok(typ.column_types.clone())
+            }
+            Get { typ, id } => {
+                if let Id::Global(_global_id) = id {
+                    if !ctx.contains_key(id) {
+                        // TODO where can we find these types
```

Comment:
They come out of the catalog, and should be populated first in HIR and the lowered into MIR. I bet someone else can be of more help!